core/domain/auth_services_test.py

```diff
@@ -26,13 +32,31 @@
 from core.platform import models
 from core.tests import test_utils
 
+from firebase_admin import auth as firebase_auth
+from typing import Dict, List, Optional, Tuple, Union, cast
 import webapp2
 
 MYPY = False
 if MYPY: # pragma: no cover
     from mypy_imports import auth_models
 
-auth_models, = models.Registry.import_models([models.Names.AUTH])
+auth_models, user_models = (
```

Comment:
Hi @cyc0825 -- thanks for your PR!

Firstly, though, we actually really don't want to duplicate the same code in multiple places (cf. the DRY principle) -- it ends up resulting in two copies of the same code, which is very hard to maintain. This code is copied from firebase_auth_services_test.py, so I just wanted to check what caused you to do that?

If you could update the PR description with the exact problem you are trying to solve, then we could probably work together to find a way to do this that does not involve duplicating code. But before I can help I need to know what lines you are trying to cover and why you cannot cover them without all this duplication.

Thanks!

Reply:
> Hi @cyc0825 -- thanks for your PR!
> 
> Firstly, though, we actually really don't want to duplicate the same code in multiple places (cf. the DRY principle) -- it ends up resulting in two copies of the same code, which is very hard to maintain. This code is copied from firebase_auth_services_test.py, so I just wanted to check what caused you to do that?
> 
> If you could update the PR description with the exact problem you are trying to solve, then we could probably work together to find a way to do this that does not involve duplicating code. But before I can help I need to know what lines you are trying to cover and why you cannot cover them without all this duplication.
> 
> Thanks!

Hi, in the test file `auth_services_test.py`, I need to cover these four functions and only the four lines in it:
<img width="629" alt="image" src="https://user-images.githubusercontent.com/30609614/206318891-6a0eaae6-1813-4071-848b-aa66f45dc9ea.png">
<img width="596" alt="image" src="https://user-images.githubusercontent.com/30609614/206318952-b51605e1-f082-401d-8534-f0fa735274eb.png">
I first try to write my own test functions, but it seems like everytime I need to create a firebase account for testing, it would give me an error below:
`google.auth.exceptions.DefaultCredentialsError: Could not automatically determine credentials. Please set GOOGLE_APPLICATION_CREDENTIALS or explicitly create credentials and re-run the application. For more information`
Later I find that these four functions all call the functions defined in `firebase_auth_services.py`, so I decide to copy paste the test cases from the `firebase_auth_services_test.py`. I wonder if there are any codes I could delete or modify in order to avoid duplicates or I could mock an account and test for its super_admin_privilage.

Reply:
Thanks for explaining. Here is what to do:

- Change the usages of firebase_auth_services in auth_services.py to platform_auth_services instead, and remove the import of firebase_auth_services.
- In your tests, stub out the relevant functions in firebase_auth_services and spy on them. Then, call the function that you're testing and assert that the relevant function in firebase_auth_services has been called. There are helper classes like CallCounter in test_utils.py that might help you with this.

I hope this helps, and gives you a way forward.

Reply:
Hi, I am still struggling how to create a firebase account and test for giving it a firebase account. I did the following:
`auth_models.UserAuthDetailsModel(id='uid', firebase_auth_id='aid').put()`
       ` USER_1 = firebase_auth.UserRecord({
            'localId': 'aid', 'email': 'email', 'disabled': False,
        })`
Without a complete FirebaseAuthServicesTestBase, after I call the function I need to cover, it kept giving me the error `ValueError: user_id=uid has no Firebase account`
This error comes from here:
<img width="629" alt="image" src="https://user-images.githubusercontent.com/30609614/206574655-3c248ac9-44b6-4fff-b998-0dbe150228d8.png">
When I check the `get_auth_id_from_user_id` function, there is an error and it seems my test file does not successfully call the function since when I add print(auth_id) like that:
<img width="749" alt="image" src="https://user-images.githubusercontent.com/30609614/206575347-5307bc09-858e-463b-9004-93b1a210baf7.png">
It does not print anything.
I cant neither debug my test file locally using breakpoint() because everytime I run the python without the scripts.run_backend_tests. It would give me an error like this:
<img width="988" alt="image" src="https://user-images.githubusercontent.com/30609614/206576743-fb3c2d27-e069-420d-b0aa-43fb493ac7cf.png">

Reply:
Update:
I find that if I create a class and put this line `USER_1 = firebase_auth.UserRecord({ 'localId': 'aid', 'email': 'email', 'disabled': False, })` into the class. I can successfully run the function above: `get_auth_id_from_user_id` and get the right auth id (which I dont understand why it is fixed this way). 

But it raise the error again: 
`raise exceptions.DefaultCredentialsError(_HELP_MESSAGE)
google.auth.exceptions.DefaultCredentialsError: Could not automatically determine credentials. Please set GOOGLE_APPLICATION_CREDENTIALS or explicitly create credentials and re-run the application. For more information, please see https://cloud.google.com/docs/authentication/getting-started
` I have no idea how to fix it.